Add `grblas.op` namespace to put all operations together.

eriknw opened the pull request:
Also, `grblas.op.numpy`, which remains lazy.

We now have `grblas.ops` and `grblas.op`.  Perhaps we should rename `grblas.ops`.

This was @jim22k's idea to help make `grblas` easier to use.

eriknw commented:
Some methods like `build` and `apply(op, right=1)` require a binaryop and don't accept monoids.  Having the monoids in `grblas.op` instead of binaryops may make using these methods more annoying.

Should we always convert monoids to binaryop when appropriate?  This should always be safe and possible to do, right?  This is similar to upgrading builtin binaryops to monoids when possible in ewise_add.

jim22k commented:
Given users can `grblas.binary.plus` and almost never need to access `grblas.ops`, lets rename it to `grblas.operators` or something like that.

And yes, auto-converting from monoid to binary or binary to monoid as required by a method seems okay to me.

eriknw commented:
- Renamed to `grblas.ops` to `grblas.operator`.
- Allow monoids any place where binaryop is required.
- Allow binaryop where monoids are required if it is part of a monoid.
- Added attributes such as `binop.monoid`, `monoid.binaryop`, and `monoid.identity`.

All this made this PR much more substantial.

Question: can we not allow using a semiring where a binaryop or monoid is expected?  For example, with ewise_add.
```python
min_plus(A | B)  # bad.  does this do min or plus?
min_plus.binary(A | B)  # better
plus(A | B)  # best

min_plus(A & B)  # bad
min_plus.monoid(A & B)  # better
min(A & B)  # best
```

eriknw commented:
I should have said I plan to disallow using semiring where it grabs the binary/monoid from it such as ewise_add.  Let me know if you prefer otherwise.

eriknw commented on grblas/operator.py:
```diff
@@ -205,10 +250,11 @@ class OpBase:
         self.name = name
         self._typed_ops = {}
         self.types = {}
+        self._anonymous = anonymous
```
Should this be private or not?

eriknw commented:
Note that we no longer include Monoids in `grblas.op`.  BinaryOps are more general w.r.t. the types they can operate on.

eriknw commented:
@jim22k one more question for you.  When users register new operators, we assign them to e.g. `grblas.binary.my_udf` if they are not anonymous.  Should we also save it to e.g. `grblas.op.my_udf`?

eriknw commented on grblas/tests/test_numpyops.py:
```diff
@@ -177,19 +177,20 @@ def test_npmonoid():
+    #                 np.array(values2, dtype=np.complex128),
+    #             ],
+    #         ]
+    #     )
```
Btw @jim22k complex monoids still don't work.  I apparently added a check in "operators.py" to avoid the segfault.

eriknw commented:
Well, this "small" PR became quite large.  Merging!
